Add support for decorating MethodViews with @Blueprint.route()

dougthor42 opened the pull request:
This updates the `Blueprint.route` decorator to support decorating `MethodView` classes.

Fixes Issue #3404.

#### Comments
+ I assume this will be part of the v2.0.0 release, so that's what I used for `versionchanged` and the changelog entry. If this is incorrect please let me know.
+ I wasn't sure where to put the tests - the code changes are part of `blueprints.py` but really only effect `views`, so I put it in `view.py`.
+ Let me know what other tests should be added.

<!--
Commit checklist:

* add tests that fail without the patch
* ensure all tests pass with ``pytest``
* add documentation to the relevant docstrings or pages
* add ``versionadded`` or ``versionchanged`` directives to relevant docstrings
* add a changelog entry if this patch changes code

Tests, coverage, and docs will be run automatically when you submit the pull
request, but running them yourself can save time.
-->

davidism commented:
Thanks for looking at this. However, I don't think I want to merge this at this time.

You can already use `MethodView` with `Blueprint` with `bp.add_url_rule()`, which corresponds to `app.add_url_rule()`. I don't want to add a special case for certain types, even if they are built-in. This also obscures the behavior of `as_view()`, which is important for `View` classes.

davidism commented:
Going to close this due to lack of response. I'm open to merging this in the future, but it needs improvement.

dougthor42 commented:
Sorry for the delay.

> I want the docs to be clear that `route` is a shortcut that doesn't pass class arguments and sets the endpoint name to the class name.

I've added a blurb for this to the docs. Let me know if you're happy with it. I also added a second example for Blueprints+MethodViews.

> apply to `add_url_rule` and `route` on both `Flask` and `Blueprint`, for any `View` subclass, not just `MethodView`

Can you elaborate on this? My naive interpretation is just:

```diff
diff --git a/src/flask/blueprints.py b/src/flask/blueprints.py
index 0e974ab2..13392174 100644
--- a/src/flask/blueprints.py
+++ b/src/flask/blueprints.py
@@ -14,6 +14,7 @@ from functools import update_wrapper
 from .helpers import _endpoint_from_view_func
 from .helpers import _PackageBoundObject
 from .views import MethodViewType
+from .views import View

 # a singleton sentinel value for parameter defaults
 _sentinel = object()
@@ -283,7 +284,7 @@ class Blueprint(_PackageBoundObject):
             endpoint = options.pop("endpoint", f.__name__)

             # Support decorating MethodView classes. See Issue #3404
-            if isinstance(f, MethodViewType):
+            if isinstance(f, (View, MethodViewType)):
                 f = f.as_view(endpoint)

             self.add_url_rule(rule, endpoint, f, **options)
```

Is that what you're talking about or am I completely misunderstanding?

dougthor42 commented:
Oh, and you can see the changes in https://github.com/dougthor42/flask/tree/blueprints-and-methodviews.
